stress: failed test in cockroach/kv/kv.test: TestTxnDBLostDeleteAnomaly

Binary: cockroach/static-tests.tar.gz sha: https://github.com/cockroachdb/cockroach/commits/f55bfc43895b6bd370112ce2af6f5d9d97fa139f

Stress build found a failed test:

```
=== RUN   TestTxnDBLostDeleteAnomaly
W160531 04:34:28.891845 gossip/gossip.go:897  not connected to cluster; use --join to specify a connected node
I160531 04:34:28.891894 storage/engine/rocksdb.go:137  opening in memory rocksdb instance
I160531 04:34:28.893474 kv/txn_correctness_test.go:525  verifying all possible histories for the "lost update (delete)" anomaly
I160531 04:34:28.894221 storage/replica_command.go:1484  range 1: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 1s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=1970-01-01 00:00:00 +0000 UTC]
E160531 04:34:28.895192 config/config.go:382  unable to determine largest reserved object ID from system config: descriptor table not found in system config of 0 values
--- FAIL: TestTxnDBLostDeleteAnomaly (0.41s)
	txn_correctness_test.go:581: ([R1(A) D2(A) W1(B-A) D2(B) C2 C1]): unexpected failure running W1(B-A): txn "kv/txn_correctness_test.go:633 txn2" id=9d887dae key="27.A" rw=true pri=0.00000005 iso=SNAPSHOT stat=PENDING epo=3 ts=0.000000000,2177 orig=0.000000000,2177 max=0.000000000,2177 wto=false rop=false failed to push "kv/txn_correctness_test.go:633 txn1" id=90a2baba key="27.B" rw=true pri=0.00000009 iso=SNAPSHOT stat=PENDING epo=0 ts=0.000000000,2171 orig=0.000000000,2171 max=0.000000000,2171 wto=false rop=false
```

Run Details:

```
0 runs so far, 0 failures, over 5s
0 runs so far, 0 failures, over 10s
0 runs so far, 0 failures, over 15s
8 runs so far, 0 failures, over 20s
16 runs so far, 0 failures, over 25s
16 runs so far, 0 failures, over 30s
16 runs so far, 0 failures, over 35s
25 runs so far, 0 failures, over 40s
32 runs so far, 0 failures, over 45s
32 runs so far, 0 failures, over 50s
32 runs so far, 0 failures, over 55s
40 runs so far, 0 failures, over 1m0s
47 runs so far, 0 failures, over 1m5s
48 runs so far, 0 failures, over 1m10s
50 runs so far, 0 failures, over 1m15s
54 runs so far, 0 failures, over 1m20s
62 runs so far, 0 failures, over 1m25s
65 runs so far, 0 failures, over 1m30s
66 runs so far, 0 failures, over 1m35s
67 runs so far, 0 failures, over 1m40s
79 runs so far, 0 failures, over 1m45s
80 runs so far, 0 failures, over 1m50s
82 runs so far, 0 failures, over 1m55s
84 runs so far, 0 failures, over 2m0s
94 runs so far, 0 failures, over 2m5s
97 runs so far, 0 failures, over 2m10s
99 runs so far, 0 failures, over 2m15s
101 runs so far, 0 failures, over 2m20s
109 runs so far, 0 failures, over 2m25s
111 runs so far, 0 failures, over 2m30s
115 runs so far, 0 failures, over 2m35s
117 runs so far, 0 failures, over 2m40s
125 runs so far, 0 failures, over 2m45s
129 runs so far, 0 failures, over 2m50s
130 runs so far, 0 failures, over 2m55s
134 runs so far, 0 failures, over 3m0s
140 runs so far, 0 failures, over 3m5s
144 runs so far, 0 failures, over 3m10s
146 runs so far, 0 failures, over 3m15s
149 runs so far, 0 failures, over 3m20s
155 runs so far, 0 failures, over 3m25s
160 runs completed, 1 failures, over 3m29s
FAIL
```
Please assign, take a look and update the issue accordingly.

--- a/kv/txn_correctness_test.go
+++ b/kv/txn_correctness_test.go
@@ -86,18 +86,20 @@ type cmd struct {
 }
 
 func (c *cmd) init(prevCmd *cmd) {
+	c.ch = nil
 	if prevCmd != nil {
+		prevCmd.ch = make(chan struct{}) // unbuffered to force synchronous handoff
 		c.prev = prevCmd.ch
 	} else {
 		c.prev = nil
 	}
-	c.ch = make(chan struct{}, 1)
 	c.debug = ""
 }
 
 func (c *cmd) execute(txn *client.Txn, t *testing.T) (string, error) {
 	if c.prev != nil {
 		<-c.prev
+		c.prev = nil
 	}
 	if log.V(2) {
 		log.Infof("executing %s", c)
@@ -105,6 +107,7 @@ func (c *cmd) execute(txn *client.Txn, t *testing.T) (string, error) {
 	err := c.fn(c, txn, t)
 	if c.ch != nil {
 		c.ch <- struct{}{}
+		c.ch = nil
 	}
 	if len(c.key) > 0 && len(c.endKey) > 0 {
 		return fmt.Sprintf("%s%%d.%%d(%s-%s)%s", c.name, c.key, c.endKey, c.debug), err
@@ -116,9 +119,14 @@ func (c *cmd) execute(txn *client.Txn, t *testing.T) (string, error) {
 }
 
 func (c *cmd) done() {
-	close(c.ch)
-	c.ch = nil
-	c.prev = nil
+	if c.prev != nil {
+		<-c.prev
+		c.prev = nil
+	}
+	if c.ch != nil {
+		c.ch <- struct{}{}
+		c.ch = nil
+	}
 	c.debug = ""
 }
 
@@ -139,12 +147,12 @@ func (c *cmd) getEndKey() []byte {
 
 func (c *cmd) String() string {
 	if len(c.key) > 0 && len(c.endKey) > 0 {
-		return fmt.Sprintf("%s%d(%s-%s)", c.name, c.txnIdx, c.key, c.endKey)
+		return fmt.Sprintf("%s%d(%s-%s)", c.name, c.txnIdx+1, c.key, c.endKey)
 	}
 	if len(c.key) > 0 {
-		return fmt.Sprintf("%s%d(%s)", c.name, c.txnIdx, c.key)
+		return fmt.Sprintf("%s%d(%s)", c.name, c.txnIdx+1, c.key)
 	}
-	return fmt.Sprintf("%s%d", c.name, c.txnIdx)
+	return fmt.Sprintf("%s%d", c.name, c.txnIdx+1)
 }
 
 // readCmd reads a value from the db and stores it in the env.
@@ -313,7 +321,7 @@ func parseHistory(txnIdx int, history string, t *testing.T) []*cmd {
 func parseHistories(histories []string, t *testing.T) [][]*cmd {
 	var results [][]*cmd
 	for i, history := range histories {
-		results = append(results, parseHistory(i+1, history, t))
+		results = append(results, parseHistory(i, history, t))
 	}
 	return results
 }
@@ -535,7 +543,7 @@ func (hv *historyVerifier) run(isolations []roachpb.IsolationType, db *client.DB
 		for _, i := range enumIso {
 			for _, h := range enumHis {
 				if err := hv.runHistory(historyIdx, p, i, h, db, t); err != nil {
-					t.Errorf("expected success, experienced %v", err)
+					t.Errorf("expected success, experienced %s", err)
 					return
 				}
 				historyIdx++
@@ -558,7 +566,7 @@ func (hv *historyVerifier) runHistory(historyIdx int, priorities []int32,
 	}
 	plannedStr := historyString(cmds)
 	if log.V(1) {
-		log.Infof("attempting iso=%v pri=%v history=%s", isolations, priorities, plannedStr)
+		log.Infof("attempting history %d: iso=%v pri=%v history=%s", historyIdx, isolations, priorities, plannedStr)
 	}
 
 	hv.actual = []string{}
@@ -568,13 +576,19 @@ func (hv *historyVerifier) runHistory(historyIdx int, priorities []int32,
 	for _, c := range cmds {
 		c.historyIdx = historyIdx
 		txnMap[c.txnIdx] = append(txnMap[c.txnIdx], c)
+		// Because the command channels are unbuffered, don't wait if the
+		// previous command is from the same transaction (and thus will be
+		// executed in a single goroutine).
+		if prev != nil && c.txnIdx == prev.txnIdx {
+			prev = nil
+		}
 		c.init(prev)
 		prev = c
 	}
 	for i, txnCmds := range txnMap {
 		go func(i int, txnCmds []*cmd) {
-			if err := hv.runTxn(i, priorities[i-1], isolations[i-1], txnCmds, db, t); err != nil {
-				t.Errorf("(%s): unexpected failure running %s: %v", cmds, cmds[i], err)
+			if err := hv.runTxn(i, priorities[i], isolations[i], txnCmds, db, t); err != nil {
+				t.Errorf("(%s): unexpected failure running %s: %s", cmds, cmds[i], err)
 			}
 		}(i, txnCmds)
 	}
@@ -653,6 +667,9 @@ func (hv *historyVerifier) runTxn(txnIdx int, priority int32,
 		for i := range cmds {
 			cmds[i].env = env
 			if err := hv.runCmd(txn, txnIdx, retry, i, cmds, t); err != nil {
+				if log.V(1) {
+					log.Infof("%s encountered error: %s", cmds[i], err)
+				}
 				return err
 			}
 		}

